csharp/src/Apache.Arrow/RecordBatch.Builder.cs

```diff
@@ -0,0 +1,149 @@
+// limitations under the License.
+
+using Apache.Arrow.Memory;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Apache.Arrow
+{
+    public partial class RecordBatch
+    {
+        public class ArrayBuilder
+        {
+            private readonly MemoryPool _memoryPool;
+
+            internal ArrayBuilder(MemoryPool memoryPool)
+            {
+                _memoryPool = memoryPool;
+            }
+
+            public BooleanArray Boolean(Action<BooleanArray.Builder> action) => Build<BooleanArray, BooleanArray.Builder>(action);
+            public Int8Array Int8(Action<Int8Array.Builder> action) => Build<Int8Array, Int8Array.Builder>(action);
+            public Int16Array Int16(Action<Int16Array.Builder> action) => Build<Int16Array, Int16Array.Builder>(action);
+            public Int32Array Int32(Action<Int32Array.Builder> action) => Build<Int32Array, Int32Array.Builder>(action);
+            public Int64Array Int64(Action<Int64Array.Builder> action) => Build<Int64Array, Int64Array.Builder>(action);
+            public UInt8Array UInt8(Action<UInt8Array.Builder> action) => Build<UInt8Array, UInt8Array.Builder>(action);
+            public UInt16Array UInt16(Action<UInt16Array.Builder> action) => Build<UInt16Array, UInt16Array.Builder>(action);
+            public UInt32Array UInt32(Action<UInt32Array.Builder> action) => Build<UInt32Array, UInt32Array.Builder>(action);
+            public UInt64Array UInt64(Action<UInt64Array.Builder> action) => Build<UInt64Array, UInt64Array.Builder>(action);
+            public FloatArray Float(Action<FloatArray.Builder> action) => Build<FloatArray, FloatArray.Builder>(action);
+            public DoubleArray Double(Action<DoubleArray.Builder> action) => Build<DoubleArray, DoubleArray.Builder>(action);
+            public Date32Array Date32(Action<Date32Array.Builder> action) => Build<Date32Array, Date32Array.Builder>(action);
+            public Date64Array Date64(Action<Date64Array.Builder> action) => Build<Date64Array, Date64Array.Builder>(action);
+            public BinaryArray Binary(Action<BinaryArray.Builder> action) => Build<BinaryArray, BinaryArray.Builder>(action);
+            public StringArray String(Action<StringArray.Builder> action) => Build<StringArray, StringArray.Builder>(action);
+
+            private TArray Build<TArray, TArrayBuilder>(Action<TArrayBuilder> action)
+                where TArray: IArrowArray
+                where TArrayBuilder: IArrowArrayBuilder<TArray>, new()
+            {
+                if (action == null)
+                {
+                    return default;
+                }
+
+                var builder = new TArrayBuilder();
+                action(builder);
+
+                return builder.Build(_memoryPool);
+            }
+        }
+
+        public class Builder
+        {
+            private readonly MemoryPool _memoryPool;
+            private readonly Schema.Builder _schema;
```

Comment:
(minor) Should this be named `_schemaBuilder`? It is a bit confusing reading the code to see `_schema`, but it really being a builder.